portability-gateway/src/main/java/org/dataportabilityproject/gateway/action/startjob/StartJobAction.java

```diff
@@ -0,0 +1,209 @@
+    // Update this job with
+    encryptAndUpdateJobWithCredentials(
+        jobId,
+        job,
+        request.getEncryptedExportAuthCredential(),
+        request.getEncryptedImportAuthCredential());
+
+    return StartJobActionResponse.create(jobId);
+  }
+
+  /**
+   * Update the data store to {@code State.CREDS_AVAILABLE} state. This indicates to the pool of
+   * workers that this job is available for processing.
+   */
+  private void updateStateToCredsAvailable(UUID jobId) {
+    PortabilityJob job = store.findJob(jobId);
+    validateJob(job);
+
+    // Set update job auth data
+    JobAuthorization jobAuthorization =
+        job.jobAuthorization().toBuilder().setState(State.CREDS_AVAILABLE).build();
+
+    job = job.toBuilder().setAndValidateJobAuthorization(jobAuthorization).build();
+    try {
+      store.updateJob(jobId, job);
+      logger.debug("Updated job {} to CREDS_AVAILABLE", jobId);
+    } catch (IOException e) {
+      throw new RuntimeException("Unable to update job", e);
+    }
+  }
+
+  /**
+   * Polls until the a public key is assigned and persisted with the job. This key will subsquently
+   * be used to encrypt credentials.
+   */
+  private PortabilityJob pollForPublicKey(UUID jobId) {
+    // Loop until the worker updates it to assigned without auth data state, e.g. at that point
+    // the worker instance key will be populated
+    // TODO: start new thread
+    // TODO: implement timeout condition
+    // TODO: Handle case where API dies while waiting
+    PortabilityJob job = store.findJob(jobId);
+    while (job == null
+        || job.jobAuthorization().state()
+            != JobAuthorization.State.CREDS_ENCRYPTION_KEY_GENERATED) {
+      logger.debug("Waiting for job {} to enter state CREDS_ENCRYPTION_KEY_GENERATED", jobId);
+      try {
+        Sleeper.DEFAULT.sleep(10000);
+      } catch (InterruptedException e) {
+        throw new RuntimeException(e);
+      }
+      job = store.findJob(jobId);
+    }
+
+    logger.debug("Got job {} in state CREDS_ENCRYPTION_KEY_GENERATED", jobId);
+
+    Preconditions.checkNotNull(
+        job.jobAuthorization().encodedPublicKey(),
+        "Expected job "
+            + jobId
+            + " to have a worker instance's public key after being assigned "
+            + "(state CREDS_ENCRYPTION_KEY_GENERATED)");
+    Preconditions.checkState(
+        job.jobAuthorization().encryptedExportAuthData() == null,
+        "Didn't expect job "
+            + jobId
+            + " to have encrypted export auth data yet in state "
+            + "CREDS_ENCRYPTION_KEY_GENERATED");
+    Preconditions.checkState(
+        job.jobAuthorization().encryptedImportAuthData() == null,
+        "Didn't expect job "
+            + jobId
+            + " to have encrypted import auth data yet in state "
+            + "CREDS_ENCRYPTION_KEY_GENERATED");
+
+    return job;
+  }
+
+  /** TODO: Determine if we need this sanity check. */
+  private void validateJob(PortabilityJob job) {
```

Comment:
same here, can all validation be done within PortabilityJob based on the job's authorization state?

Reply:
Same